src/main/c/output.c

```diff
@@ -605,3 +605,73 @@ jobject output_convert_schema_checked(JNIEnv *env, const ddwaf_object *obj)
     JNI(DeleteLocalRef, ret);
     return NULL;
 }
+
+jobject convert_ddwaf_object_to_jobject(JNIEnv *env, const ddwaf_object *obj)
+{
+    switch (obj->type) {
+        // TODO: Implement missing types, when needed
+        case DDWAF_OBJ_INVALID:
+        case DDWAF_OBJ_SIGNED:
+        case DDWAF_OBJ_UNSIGNED:
+        case DDWAF_OBJ_FLOAT:
+        case DDWAF_OBJ_BOOL:
+            return NULL;
+        case DDWAF_OBJ_STRING:
+            return java_utf8_to_jstring_checked(env, obj->stringValue, obj->nbEntries);
+        case DDWAF_OBJ_ARRAY: {
+            jobject ret = java_meth_call(env, &_array_list_init, NULL, (jint) obj->nbEntries);
+            if (ret == NULL) {
+                return NULL;
+            }
+
+            for (uint64_t i = 0; i < obj->nbEntries; i++) {
+                jobject elem = convert_ddwaf_object_to_jobject(env, &obj->array[i]);
+                if (elem == NULL) {
+                    JNI(DeleteLocalRef, ret);
+                    return NULL;
+                }
+
+                java_meth_call(env, &_array_list_add, ret, elem);
+                JNI(DeleteLocalRef, elem);
+                if (JNI(ExceptionCheck)) {
+                    JNI(DeleteLocalRef, ret);
+                    return NULL;
+                }
+            }
+
+            return ret;
+        }
+        case DDWAF_OBJ_MAP: {
+            jobject ret = java_meth_call(env, &_linked_hm_init, NULL);
+            if (ret == NULL) {
+                return NULL;
+            }
+
+            for (uint64_t i = 0; i < obj->nbEntries; i++) {
+                ddwaf_object *elem = &obj->array[i];
+                jstring key = java_utf8_to_jstring_checked(env, elem->parameterName, elem->parameterNameLength);
+                if (JNI(ExceptionCheck)) {
+                    JNI(DeleteLocalRef, ret);
+                    return NULL;
+                }
+
+                jobject value = convert_ddwaf_object_to_jobject(env, elem);
+                if (value == NULL) {
+                    JNI(DeleteLocalRef, key);
+                    continue;
```

Comment:
You shouldn't be continuing here. Presumably there is an exception pending already, which all cause all subsequent calls to fail until it's cleared. You should rather give up like you do for the keys.

Reply:
Fixed